[Feature] Add support datacache.partition_duration for mv

stonechen111 opened the pull request:
Why I'm doing:
Materialized view doesn't support to set datacache.partition_duration so far.

What I'm doing:
Add support datacache.partition_duration property for materialized view to
control the validity time of its datacache.

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

kevincai commented on fe/fe-core/src/test/java/com/starrocks/analysis/CreateMaterializedViewTest.java:
```diff
@@ -3960,4 +3960,21 @@ public void testCreateMvWithUnsupportedStr2date() {
+                    "\"datacache.partition_duration\" = \"6 day\"" +
+                    ") " +
+                    "as select str2date(d, '%m-%d-%Y') ss, a, b, c from jdbc0.partitioned_db0.tbl1;";
+            Assert.assertThrows("Materialized view partition function date_trunc check failed",
```
why just add an invalid case, should cover the success case, check if the partition_duration is set correctly.

UT is not just for code coverage, should be always function oriented.

stonechen111 replied:
OK

stonechen111 committed:
[Feature] Add support datacache.partition_duration for mv.

Add support datacache.partition_duration property for materialized view to
control the validity time of its datacache.

Signed-off-by: stonechen111 <chenshi@starrocks.com>

stonechen111 commented on fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:
```diff
@@ -3433,6 +3433,12 @@ private DataProperty analyzeMVDataProperties(Database db,
                 setLakeStorageInfo(materializedView, storageVolumeId, properties);
             }
 
+            // datacache.partition_duration
```
> what's the behavior if the property is not set? will it have a default value or something else, can this behavior be covered by some UTs as well?

If this parameter is not set, it will have no default value and tableProperty.dataCachePartitionDuration will remain null. The above content is covered in OlapTableTest.testNullDataCachePartitionDuration.

kevincai replied:
> OlapTableTest.testNullDataCachePartitionDuration

does it also cover mv case? lke if mv is not provided with datacache partition duration, it will be null and the check shouldCache to the partition are always to `false` result?

stonechen111 replied:
> > what's the behavior if the property is not set? will it have a default value or something else, can this behavior be covered by some UTs as well?
> 
> If this parameter is not set, it will have no default value and tableProperty.dataCachePartitionDuration will remain null. The above content is covered in OlapTableTest.testNullDataCachePartitionDuration.

MaterializedView is a subclass of OlapTable. Mv's OlapTable.tableProperty.dataCachePartitionDuration is set in analyzeMVDataProperties. All entries for using datacache are the same: olapTable.isEnableFillDataCache (in OlapScanNode.addScanRangeLocations)

sduzh commented:
@wyb PTAL

wyb commented on fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:
```diff
@@ -3433,6 +3433,12 @@ private DataProperty analyzeMVDataProperties(Database db,
             }
 
+            // datacache.partition_duration
+            if (properties.containsKey(PropertyAnalyzer.PROPERTIES_DATACACHE_PARTITION_DURATION)) {
```
better move it into `if (materializedView.isCloudNativeMaterializedView()) {}`

stonechen111 committed:
limit datacache.partition_duration to lake mv

Signed-off-by: stonechen111 <chenshi@starrocks.com>

sduzh commented:
@Mergifyio backport branch-3.2

sduzh commented:
@Mergifyio backport branch-3.1
